File: CombatCore/Fixed.cs
namespace CombatCore;

/// <summary>
/// Q16.16 fixed-point number. This exists for exactly one reason: float and double
/// arithmetic is NOT guaranteed to produce bit-identical results across different
/// CPUs, JIT versions, or between .NET (server) and a browser's WASM/JS runtime
/// (client). A single ULP of divergence in a position calculation compounds every
/// tick and eventually desyncs client prediction from server truth.
///
/// Every operation here is integer-only. int32 overflow/wraparound behavior is
/// identical in C# (checked/unchecked) and JavaScript (via `| 0` and friends), which
/// is what makes it possible to port this type to TypeScript with byte-for-byte
/// matching behavior - see client/src/sim/Fixed.ts, which is a deliberate mirror
/// of this file, not just "similar logic".
///
/// Range: with 16 fractional bits, integer range is roughly +/-32,768 with a
/// resolution of 1/65536 (~0.0000153). That is enormous headroom for a fighting
/// game arena measured in a handful of meters - we will never get near overflow
/// on position values, only potentially on intermediate multiplication, which is
/// why multiply/divide widen to long/int64 before shifting back down.
/// </summary>
public readonly struct Fixed : IEquatable<Fixed>, IComparable<Fixed>
{
    public const int FractionalBits = 16;
    public const int RawOne = 1 << FractionalBits;

    public readonly int Raw;

    private Fixed(int raw) => Raw = raw;

    public static readonly Fixed Zero = new(0);
    public static readonly Fixed One = new(RawOne);
    public static readonly Fixed Half = new(RawOne / 2);
    public static readonly Fixed MinusOne = new(-RawOne);

    public static Fixed FromRaw(int raw) => new(raw);
    public static Fixed FromInt(int value) => new(value << FractionalBits);

    /// <summary>
    /// ONLY for authoring move data / tests at design time (e.g. "this hitbox is
    /// 0.75 units wide"). Never call this inside the simulation tick itself -
    /// float-to-fixed conversion is a one-time, load-time operation, not a
    /// per-frame one, precisely so no float ever enters the hot path.
    /// </summary>
    public static Fixed FromFloat(float value) => new((int)MathF.Round(value * RawOne));

    public float ToFloat() => Raw / (float)RawOne;
    public int ToIntFloor() => Raw >> FractionalBits;

    public static Fixed operator +(Fixed a, Fixed b) => new(a.Raw + b.Raw);
    public static Fixed operator -(Fixed a, Fixed b) => new(a.Raw - b.Raw);
    public static Fixed operator -(Fixed a) => new(-a.Raw);

    public static Fixed operator *(Fixed a, Fixed b) =>
        new((int)(((long)a.Raw * b.Raw) >> FractionalBits));

    public static Fixed operator /(Fixed a, Fixed b) =>
        new((int)(((long)a.Raw << FractionalBits) / b.Raw));

    public static Fixed operator *(Fixed a, int scalar) => new(a.Raw * scalar);
    public static Fixed operator /(Fixed a, int scalar) => new(a.Raw / scalar);

    public static bool operator ==(Fixed a, Fixed b) => a.Raw == b.Raw;
    public static bool operator !=(Fixed a, Fixed b) => a.Raw != b.Raw;
    public static bool operator <(Fixed a, Fixed b) => a.Raw < b.Raw;
    public static bool operator >(Fixed a, Fixed b) => a.Raw > b.Raw;
    public static bool operator <=(Fixed a, Fixed b) => a.Raw <= b.Raw;
    public static bool operator >=(Fixed a, Fixed b) => a.Raw >= b.Raw;

    public static Fixed Abs(Fixed a) => a.Raw < 0 ? new Fixed(-a.Raw) : a;
    public static Fixed Min(Fixed a, Fixed b) => a.Raw < b.Raw ? a : b;
    public static Fixed Max(Fixed a, Fixed b) => a.Raw > b.Raw ? a : b;
    public static Fixed Clamp(Fixed v, Fixed lo, Fixed hi) => Max(lo, Min(hi, v));

    /// <summary>
    /// Deterministic integer square root (bit-by-bit / "digit-by-digit" method).
    /// Hardware sqrt instructions and Math.Sqrt are NOT guaranteed bit-identical
    /// across platforms, so anything gameplay-relevant that needs a magnitude
    /// (e.g. normalizing a movement vector) must go through this instead.
    /// Operates on the raw fixed value scaled up so the result keeps full
    /// fractional precision.
    /// </summary>
    public static Fixed Sqrt(Fixed value)
    {
        if (value.Raw <= 0) return Zero;

        // We want sqrt(value.Raw / RawOne) * RawOne = sqrt(value.Raw * RawOne)
        long operand = (long)value.Raw << FractionalBits;
        long result = 0;
        long bit = 1L << 62;

        while (bit > operand) bit >>= 2;

        while (bit != 0)
        {
            if (operand >= result + bit)
            {
                operand -= result + bit;
                result = (result >> 1) + bit;
            }
            else
            {
                result >>= 1;
            }
            bit >>= 2;
        }

        return new Fixed((int)result);
    }

    public bool Equals(Fixed other) => Raw == other.Raw;
    public override bool Equals(object? obj) => obj is Fixed f && Equals(f);
    public override int GetHashCode() => Raw;
    public int CompareTo(Fixed other) => Raw.CompareTo(other.Raw);
    public override string ToString() => ToFloat().ToString("F4");
}

File: Client/src/sim/Fixed.ts
/**
 * Deliberate mirror of CombatCore/Fixed.cs. Every operation here must produce
 * the exact same `raw` int32 result as its C# counterpart given the same
 * inputs - that is the entire contract this file exists to satisfy, verified
 * by src/parity.test.ts against golden replays recorded from the real C# sim.
 *
 * Two things make this trickier than a typical TS port:
 *   1. JS's native `<<`/`>>`/`|` bitwise operators coerce to *32-bit* ints,
 *      which is actually what we want for the final result, but C#'s
 *      multiply/divide/sqrt widen to 64-bit `long` for the INTERMEDIATE
 *      calculation before truncating back down - plain JS `number` can't
 *      hold a 64-bit integer exactly past 2^53, so those three operations use
 *      BigInt for the intermediate math and truncate to int32 only at the end
 *      (via BigInt.asIntN(32, ...), which matches C#'s `(int)` cast on a long).
 *   2. Plain `+`/`-` on two in-range int32 values won't overflow in JS the
 *      way unchecked C# `int` arithmetic silently wraps around, so `| 0` is
 *      applied to force the same 32-bit wraparound behavior even though it
 *      should never matter at gameplay-realistic position values.
 */
export class Fixed {
  static readonly FRACTIONAL_BITS = 16;
  static readonly RAW_ONE = 1 << 16;

  readonly raw: number;

  private constructor(raw: number) {
    this.raw = raw | 0;
  }

  static fromRaw(raw: number): Fixed {
    return new Fixed(raw);
  }

  static fromInt(value: number): Fixed {
    return new Fixed(value << Fixed.FRACTIONAL_BITS);
  }

  /** Design-time / test-authoring only - never call this inside the sim tick. */
  static fromFloat(value: number): Fixed {
    return new Fixed(Math.round(value * Fixed.RAW_ONE));
  }

  toFloat(): number {
    return this.raw / Fixed.RAW_ONE;
  }

  static readonly ZERO = Fixed.fromRaw(0);
  static readonly ONE = Fixed.fromRaw(Fixed.RAW_ONE);
  static readonly HALF = Fixed.fromRaw(Fixed.RAW_ONE / 2);

  static add(a: Fixed, b: Fixed): Fixed {
    return new Fixed((a.raw + b.raw) | 0);
  }

  static sub(a: Fixed, b: Fixed): Fixed {
    return new Fixed((a.raw - b.raw) | 0);
  }

  static neg(a: Fixed): Fixed {
    return new Fixed((-a.raw) | 0);
  }

  static mul(a: Fixed, b: Fixed): Fixed {
    const product = (BigInt(a.raw) * BigInt(b.raw)) >> BigInt(Fixed.FRACTIONAL_BITS);
    return new Fixed(Number(BigInt.asIntN(32, product)));
  }

  static div(a: Fixed, b: Fixed): Fixed {
    const numerator = BigInt(a.raw) << BigInt(Fixed.FRACTIONAL_BITS);
    const quotient = numerator / BigInt(b.raw); // BigInt division truncates toward zero, matching C# long division
    return new Fixed(Number(BigInt.asIntN(32, quotient)));
  }

  static mulInt(a: Fixed, scalar: number): Fixed {
    return new Fixed(Math.imul(a.raw, scalar));
  }

  static abs(a: Fixed): Fixed {
    return a.raw < 0 ? new Fixed((-a.raw) | 0) : a;
  }

  static min(a: Fixed, b: Fixed): Fixed {
    return a.raw < b.raw ? a : b;
  }

  static max(a: Fixed, b: Fixed): Fixed {
    return a.raw > b.raw ? a : b;
  }

  static clamp(v: Fixed, lo: Fixed, hi: Fixed): Fixed {
    return Fixed.max(lo, Fixed.min(hi, v));
  }

  /** Deterministic bit-by-bit integer sqrt - see Fixed.cs for the matching C# version. */
  static sqrt(value: Fixed): Fixed {
    if (value.raw <= 0) return Fixed.ZERO;

    let operand = BigInt(value.raw) << BigInt(Fixed.FRACTIONAL_BITS);
    let result = 0n;
    let bit = 1n << 62n;

    while (bit > operand) bit >>= 2n;

    while (bit !== 0n) {
      if (operand >= result + bit) {
        operand -= result + bit;
        result = (result >> 1n) + bit;
      } else {
        result >>= 1n;
      }
      bit >>= 2n;
    }

    return new Fixed(Number(BigInt.asIntN(32, result)));
  }

  equals(other: Fixed): boolean {
    return this.raw === other.raw;
  }

  compare(other: Fixed): number {
    return this.raw - other.raw;
  }

  toString(): string {
    return this.toFloat().toFixed(4);
  }
}
